src/components/views/rooms/wysiwyg_composer/hooks/useSuggestion.ts

```diff
@@ -124,69 +213,129 @@ export const processCommand = (
- * if so
- * TODO expand this to handle mentions
+ * Given some text content from a node and the cursor position, search through the content
+ * to find a mention or a command. If there is one present, return a slice of the content
+ * from the special character to the end of the input
  *
- * @param editorRef - ref to the composer
- * @param suggestion - the current suggestion state
- * @param setSuggestion - the setter for the suggestion state
+ * @param text - the text content of a node
+ * @param offset - the current cursor offset position
+ * @returns an empty string if no mention or command is found, otherwise the mention/command substring with it's start offset
  */
-export const processSelectionChange = (
-    editorRef: React.RefObject<HTMLDivElement>,
-    suggestion: SuggestionState,
-    setSuggestion: React.Dispatch<React.SetStateAction<SuggestionState>>,
-): void => {
-    const selection = document.getSelection();
+export function findSuggestionInText(
+    text: string,
+    offset: number,
+): { text: string; startOffset: number; endOffset: number } | null {
+    // Return null early if the offset is outside the content
+    if (offset < 0 || offset > text.length) {
+        return null;
+    }
 
-    // return early if we do not have a current editor ref with a cursor selection inside a text node
-    if (
-        editorRef.current === null ||
-        selection === null ||
-        !selection.isCollapsed ||
-        selection.anchorNode?.nodeName !== "#text"
-    ) {
-        return;
+    // A cursor offset lies between two characters, so make a cursor offset correspond to
+    // a character index in the text string
+    let startCharIndex = offset - 1;
+    let endCharIndex = offset;
+
+    while (shouldDecrementStartIndex(text, startCharIndex)) {
+        // Special case - if we hit some whitespace, return null. This is to catch cases
+        // where user types a special character then whitespace
+        if (/\s/.test(text[startCharIndex])) {
+            return null;
+        }
+        startCharIndex--;
     }
 
-    // here we have established that both anchor and focus nodes in the selection are
-    // the same node, so rename to `currentNode` for later use
-    const { anchorNode: currentNode } = selection;
+    while (shouldIncrementEndIndex(text, endCharIndex)) {
+        endCharIndex++;
+    }
 
-    // first check is that the text node is the first text node of the editor, as adding paragraphs can result
-    // in nested <p> tags inside the editor <div>
-    const firstTextNode = document.createNodeIterator(editorRef.current, NodeFilter.SHOW_TEXT).nextNode();
+    // We have looped throught the text in both directions from the current cursor position
+    // whilst looking for special characters.
+    // We do not have a command or mention if:
+    // - the start or ending indices are outside the string
+    if (startCharIndex < 0 || endCharIndex > text.length) return null;
```

Comment:
I'm struggling to understand why this condition is correct.

My understanding of the logic above is:

* `startCharIndex` will be -1 if the first character is a `/` or `@`
* `endCharIndex` will be `text.length + 1` if there is no whitespace after the cursor.

Is it really correct that we do not make a suggestion in these cases?